keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,767 @@
+
+```
+<<[UNRESOLVED optional-vs-required ]>>
+
+If we wanted to be able to have both "optionally dual-stack" and
+"mandatory dual-stack" we could have separate "optional ipv4/6" and
+"required ipv4/6" flags, but that seems to result in a lot of
+complication for not much benefit.
+
+<<[/UNRESOLVED]>>
+```
+
+If `--node-ips` is not specified but `--node-ip` is, then `--node-ips`
+is defaulted to an equivalent value as follows:
+
+  - If `--node-ip` is `0.0.0.0` then `--node-ips` is set to `ipv4`.
+
+  - If `--node-ip` is `::` then `--node-ips` is set to `ipv6`.
+
+  - Otherwise (when `--node-ip` is a specific IP address) `--node-ips`
+    is set to the same value as `--node-ip`.
+
+It is an error to pass both `--node-ip` and `--node-ips`.
+
+As discussed under [Implementation](#implementation),
+`Node.Status.Addresses` will then be set correctly to reflect this.
+Once `Node.Status.Addresses` has been set, the rule for actually
+finding the Primary Node IP is still the same as it used to be; we
+will just adjust `Node.Status.Addresses` to ensure that someone
+following the traditional rule will find the IP we want them to.
+
+There is now also a "Secondary Node IP", which uses the same rule as
+for Primary Node IP, but only matching IPs of the opposite IP family.
+(Thus, the Secondary Node IP is the first `InternalIP` address of the
+opposite IP family from the Primary Node IP, or the first such
+`ExternalIP` if there is no such `InternalIP`.)
+
+#### Updated Use of `Node.Status.Addresses`
+
+When setting Pod status, kubelet will continue to use the Primary Node
+IP for `Pod.Status.HostIP`.
+
+In clusters with the `IPv6DualStack` feature gate enabled (or
+post-dual-stack GA), host-network Pods will now get both the Primary
+Node IP and the Secondary Node IP as `Pod.Status.PodIPs` (allowing
+host-network Pods to be endpoints of both IPv4 and IPv6 Services).
+
+Some methods in `utilnode` such as `GetNodeHostIP` and
+`GetPreferredNodeAddress` may be updated to have dual-stack versions
+that return multiple IPs. The existing versions are still useful on
+their own though.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+TODO
+
+## Design Details
+
+### Implementation
+
+As before, when using an external cloud provider, the external
+provider deals with setting `Node.Status.Addresses`, and when using a
+built-in cloud provider or bare metal, kubelet deals with setting it.
+
+#### With a Cloud Provider (External or Built-In)
+
+When using an external cloud provider, kubelet will pass the value of
+`--node-ips` (explicit or defaulted) to the provider via a Node
+annotation. (If the provider finds that kubelet has not set the
```

Comment:
Hm... putting it in a field in node feels weird to me since it's not meant for general consumption... it's just for communication between kubelet and the cloud provider.

Is there anything specifically "not great" about `provided-node-ip`?

The other possibility would be to have the cloud provider no longer set `Node.Status.Addresses` itself, and instead return the provisional list of node IPs to kubelet via some other means, and then in all cases (external cloud, internal cloud, bare metal), kubelet would be the one responsible for fixing up the address list and setting it on the Node. Then it doesn't need to communicate `--node-ips` to anyone else, and that also gets rid of the need to share `PatchNodeStatus`, `ParseNodeIPs`, etc too.

But in that case it's not clear how the cloud provider would return the list of addresses... I guess it could create `CloudNode` objects or something maybe, and kubelet could use those?

But also, in this scenario, the cloud provider needs to know whether it's working with an old kubelet (which expects the cloud provider to set the node addresses) or a new one (which doesn't), so you'd probably still need a node annotation anyway...

Reply:
> Is there anything specifically "not great" about provided-node-ip?

The provided-node-ip annotation only [ensures one of the provided NodeAddresses](https://github.com/kubernetes/kubernetes/blob/46d522d0c8e94edb6d29606a28785e35259badd7/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L591-L604) from the cloud provider matches this IP. It doesn't guarantee anything about the type (InternalIP or ExternalIP) or the order. 

And if the cloud provider doesn't return any NodeAddresses that match the IP in the annotation, then node initialization fails which ends up being error prone and rigid since you still can't override the node address from the cloud provider.

Reply:
Well, sure, but that has nothing to do with it being an annotation rather than a Node field. If the original creators of that annotation had wanted the external cloud provider to interpret it differently, they could have required different behavior.

The defined behavior of the new proposed annotation is that kubelet uses it to pass the value of `--node-ips` to the external cloud provider, and then the external cloud provider interprets that value in exactly the same way that kubelet interprets `--node-ips` for built-in cloud providers and bare metal. (And ideally, they should be using the same code as kubelet uses. We just need to figure out where to put it.)

The KEP doesn't currenty talk about what happens if the cloud provider can't match `--node-ips`, but I wrote most of the code to implement an earlier revision of this KEP, and the behavior I went with is that the cloud provider logs an error and then just writes its entire list of available node addresses to `Node.Status.Addresses`. Then when kubelet sees that update, it can see that the cloud provider has written out a list of addresses that don't conform to `--node-ips`, so it knows that this means the requested IPs were unavailable, so it will exit with an error, telling the user what IPs are available.